lib/proxy/ObjectCallProxy.js

```diff
@@ -0,0 +1,116 @@
+var _ = require('lodash');
+
+/**
+ * Create a new ObjectCallProxy.
+ *
+ * @class
+ */
+function ObjectCallProxy (source) {
+  this.source = source;
+  this.attached = {};
+  this.pending = [];
+}
+
+/**
+ * Restore all of the attached methods.
+ *
+ * @private
+ */
+ObjectCallProxy.prototype._restore = function () {
+  for (var name in this.attached) {
+    this.source[name] = this.attached[name];
+    delete this.attached[name];
+  }
+};
+
+/**
+ * Release all of the pending callbacks.
+ *
+ * @private
+ * @param {Error} [err] - Error to release all callbacks with (optional).
+ */
+ObjectCallProxy.prototype._release = function (err) {
+  var source = this.source;
+  var pending = this.pending;
+
+  this.pending = [];
+
+  pending.forEach(function (call) {
+    var args = call.args,
+        callback = call.callback,
+        fn = call.fn;
+
+    args[args.length - 1] = function () {
+      var subArgs = err ? [err] : Array.prototype.slice.call(arguments);
+      callback.apply(null, subArgs);
+    };
+
+    fn.apply(source, args);
+  });
+};
+
+/**
+ * Attach onto the source object and intercept all calling
+ * methods with callbacks (where last argument is function).
+ *
+ * @param fn [predicateFn] - Predicate to filter methods by (optional).
+ */
+ObjectCallProxy.prototype.attach = function (predicateFn) {
+  var outerScope = this;
+  var source = outerScope.source;
```

Comment:
`outerScope` is pretty generic and in this case actually *this* scope, and not the outer. Instead, I would declare two variables here, `attached` and `pending`:

```javascript
var attached = this.attached;
var pending = this.pending;
var source = this.source;
...
```

Then I replace the use of `outerScope.pending` to only use `pending` etc.

What do you think?